engines/cegar_values.cpp

```diff
@@ -0,0 +1,309 @@
+ProverResult CegarValues<Prover_T>::check_until(int k)
+{
+  initialize();
+
+  ProverResult res = ProverResult::FALSE;
+  while (res == ProverResult::FALSE) {
+    // need to call parent's check_until in case it
+    // is another cegar loop rather than an engine
+    res = super::check_until(k);
+
+    if (res == ProverResult::FALSE) {
+      if (!cegar_refine()) {
+        return ProverResult::FALSE;
+      }
+    }
+  }
+
+  if (res == ProverResult::TRUE && super::invar_) {
+    // update the invariant
+    UnorderedTermMap super_to_vals;
+    for (const auto & elem : to_vals_) {
+      super_to_vals[from_cegval_solver_.transfer_term(elem.first)] =
+          from_cegval_solver_.transfer_term(elem.second);
+    }
+    super::invar_ = super::solver_->substitute(super::invar_, super_to_vals);
+  }
+
+  return res;
+}
+
+template <class Prover_T>
+void CegarValues<Prover_T>::initialize()
+{
+  if (super::initialized_) {
+    return;
+  }
+
+  // specify which cegar_abstract in case
+  // we're inheriting from another cegar algorithm
+  CegarValues::cegar_abstract();
+  super::initialize();
+
+  // update local version of ts over fresh solver
+  cegval_ts_ = TransitionSystem(prover_ts_, to_cegval_solver_);
+  // update cache
+  UnorderedTermMap & cache = from_cegval_solver_.get_cache();
+  Term nv;
+  for (const auto & sv : prover_ts_.statevars()) {
+    nv = prover_ts_.next(sv);
+    cache[to_cegval_solver_.transfer_term(sv)] = sv;
+    cache[to_cegval_solver_.transfer_term(nv)] = nv;
+  }
+
+  for (const auto & iv : prover_ts_.inputvars()) {
+    cache[to_cegval_solver_.transfer_term(iv)] = iv;
+  }
+
+  // create labels for each abstract value
+  Sort boolsort = cegval_solver_->make_sort(BOOL);
+  Term lbl;
+  for (const auto & elem : to_vals_) {
+    lbl = cegval_solver_->make_symbol("__assump_" + elem.second->to_string(),
+                                      boolsort);
+    cegval_labels_[elem.first] = lbl;
+  }
+}
+
+template <class Prover_T>
+void CegarValues<Prover_T>::cegar_abstract()
+{
+  UnorderedTermMap prover_to_vals;
+  ValueAbstractor va(prover_ts_, prover_to_vals);
+
+  // now update with abstraction
+  if (prover_ts_.is_functional()) {
+    // add variables
+    for (const auto & sv : conc_ts_.statevars()) {
+      prover_ts_.add_statevar(sv, conc_ts_.next(sv));
+    }
+
+    for (const auto & iv : conc_ts_.inputvars()) {
+      prover_ts_.add_inputvar(iv);
+    }
+
+    Term init = prover_ts_.init();
+    prover_ts_.set_init(va.visit(init));
+
+    // state updates
+    for (auto elem : conc_ts_.state_updates()) {
+      prover_ts_.assign_next(elem.first, va.visit(elem.second));
+    }
+
+    for (auto con : conc_ts_.constraints()) {
+      // NOTE: there should be better infrastructure for re-adding constraints
+      // currently have to avoid re-adding the next version
+      if (conc_ts_.no_next(con)) {
+        prover_ts_.add_constraint(va.visit(con));
+      }
+    }
+  } else {
+    prover_ts_ = conc_ts_;
```

Comment:
prover_ts_ is pointing to conc_ts_. Is it intentional?

Reply:
It is not clear to me, what is happening here?

Reply:
Ahh I'll leave a comment, it's a copy constructor so I'm just starting the `prover_ts_` with all the same behavior as `conc_ts_` before abstracting it.

Reply:
It's an easy way to avoid manually copying over all the state variables etc... (although I had to do that in the functional case anyway)

Reply:
I see. So, va will have the old ts reference. Right?